pennylane/_grad.py

```diff
@@ -181,3 +182,217 @@ def _jacobian_function(*args, **kwargs):
+
+    For :math:`i = j`:
+
+    .. math::
+        \frac{\partial^2 f(x)}{\partial x_i^2} \approx
+        \frac{f(x_i + \delta) - 2 f(x) + f(x_i - \delta)}{\delta^2},
+
+    and for :math:`i \neq j`:
+
+    .. math::
+        \frac{\partial^2 f(x)}{\partial x_i \partial x_j} \approx
+        \frac{f(x_i + \delta/2, x_j + \delta/2) - f(x_i - \delta/2, x_j + \delta/2)
+        - f(x_i + \delta/2, x_j - \delta/2) + f(x_i - \delta/2, x_j - \delta/2)}
+        {\delta^2}.
+
+    Args:
+        f (function): function with signature ``f(*args, **kwargs)``
+        N (int): specifies the order of the finite difference approximation
+        argnum (int): the argument of function ``f`` to differentiate
+        idx (list[int]): If argument ``args[argnum]`` is an array, `idx`` can be used
+            to specify the indices of the argument ``argnum`` to differentiate.
+            For ``N=1`` it can be given to specify the gradient components to be computed.
+            For example, for function ``f(x, y, z)``, ``argnum=1``, ``idx=[3, 2]``the returned
+            function will differentiate ``f`` with respect to elements ``3`` and ``2`` of
+            argument ``y``. For ``N=2``, it specifies the indices ``i, j`` of the variables
+            involved in the second-order derivative
+            :math:`\frac{\partial^2 f(x, y, z)}{\partial y_i \partial y_j}`.
+        delta (float): step size used to evaluate the finite differences
+
+    Returns:
+        function: the function to compute the gradient (``N=1``) or the
+        second-order derivative (``N=2``) of the input function ``f`` with respect
+        to the arguments in ``argnum``
+
+    **Examples**
+
+    >>> def f(x, y):
+    ...     return np.sin(y[0])*np.sin(y[1]) + np.exp(y[2]) - x**-3
+
+    >>> (x, y) = (0.376, np.array([1.975, 0.33, -0.4]))
+
+    >>> gradient = qml.finite_diff(f, argnum=1)
+    >>> print(gradient(x, y))
+    [-0.12744129189670161 0.869802723370583 0.6703228390396276]
+
+    >>> derivative = qml.finite_diff(f, argnum=1, idx=[1])
+    >>> print(derivative(x, y)[1])
+    0.869802723370583
+
+    >>> second_derivative = qml.finite_diff(f, N=2, argnum=1, idx=[0, 1])
+    >>> print(second_derivative(x, y))
+    -0.37206279888124527
+    """
+
+    if not callable(f):
+        error_message = "{} object is not callable. \n" "'f' should be a callable function".format(
+            type(f)
+        )
+        raise TypeError(error_message)
+
+    if delta <= 0.0:
+        raise ValueError(
+            "The value of the step size 'delta' has to be greater than 0; got {}".format(delta)
+        )
+
+    if N == 1:
+        return partial(_fd_first_order_centered, f, argnum, delta, idx=idx)
+
+    if N == 2:
+        return partial(_fd_second_order_centered, f, argnum, delta, idx=idx)
+
+    raise ValueError(
+        "At present, finite-difference approximations are supported up to second-order."
+        " The value of 'N' can be 1 or 2; got {}".format(N)
+    )
```

Comment:
Removing the ``if`` here fixed the CodeFactor issue (which makes sense), thanks @josh146 !